### PR TITLE
Apeswap - Large LP Deposits/Withdrawals

### PR DESCRIPTION
**Pull request by Vxatz**
# Large Apeswap LP Deposits/Withdrawals bot

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [x] New Agent
- [ ] Fix an agent error
- [ ] Update an agent logic/test/documentation
- [ ] Code style update (formatting, renaming)
- [ ] Other (please describe):

## New Agent PR readiness

- [x] `npm install` command works (i.e. package-lock is updated properly)
- [x] The `name` and `description` fields of `package.json` describe the agent properly
- [x] All the libraries listed in `package.json` are being used
- [x] All the development libraries are in `devDependencies` section of `package.json`
- [x] The agent README is updated
- [x] The agent README sufficiently describes what the agent does
- [x] All the alerts that the agent emits are listed in the README
- [ ] The txn/block examples listed in the README generate the expected findings
- [x] Code is easy to follow and/or properly commented
- [x] Agent contains tests
- [x] All tests pass when running `npm test`
- [x] Tests are not making real network calls (i.e. network interaction is properly mocked)
- [x] Tests cover all the paths (i.e. negative & positive test cases)
- [x] When the tests complete, no warning/errors are shown
- [ ] `npm start` works without errors when executed using a jsonRpcUrl for each of the networks specified in the README
- [x] The networks specified in the README and the networks are listed in the `package.json` are the same
- [x] The code does what the README describes
- [x] Network calls are properly being cached (if duplicated calls can occur)
- [x] A performance review has been done (e.g. `Promise.all` is used whenever possible)
- [x] The findings fields are filled out correctly
- [x] Findings `protocol` is specified (if the agent is for a specific one)
- [x] `metadata` of the findings contains all the information needed.
- [x] The SDK methods are being used appropriately (e.g. `filterLog`, `filterFunction`, `getEthersProvider`)
- [x] `ABI`s match the contracts used / `args` used from `LogDescriptions` exists in the `ABI`s
- [x] The agent follows Forta recommended [best practices](https://docs.forta.network/en/latest/best-practices/)
- [x] Agent can recover from failed error calls when they can fail (e.g. errors are being caught on calls that can fail)
- [x] `repository` is added in the `package.json`

## Comments (optional)

Please keep in mind the `ERROR: TypeError: Cannot read properties of null (reading 'hash')` when fetching some blocks in case it is a bot's problem.

### CODE REVIEW

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/package.json [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    "@types/lru-cache": "^7.6.1",
+    "jest": "^27.0.6",
+    "jest-when": "^3.5.1",
+    "forta-agent-tools": "^2.0.5",
+    "nodemon": "^2.0.8",
+    "prettier": "^2.6.0",
+    "ts-jest": "^27.0.3",
+    "typescript": "^4.3.4"
+  },
+  "repository": "https://github.com/NethermindEth/Forta-Agents/Apeswap-Bots/Large-LP-Deposits-Withdrawals"
```
Comment (e1Ru1o) [inline Apeswap-Bots/Large-LP-Deposits-Withdrawals/package.json:41]:
```suggestion
  "repository": "https://github.com/NethermindEth/Forta-Agents/tree/main/Apeswap-Bots/Large-LP-Deposits-Withdrawals"```

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/agent.spec.ts [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+          token0,
+          amount0,
+          token1,
+          amount1,
+        },
+      });
+  }
+};
+
+describe("Apeswap Large LP Deposit/Withdrawl bot test suite", () => {
```
Comment (e1Ru1o) [inline Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/agent.spec.ts:127]:
```suggestion
describe("Apeswap Large LP Deposit/Withdrawal bot test suite", () => {
```

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/pool.fetcher.spec.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      mockProvider.addCallTo(pool, block, utils.FUNCTIONS_IFACE, "token1", {
+        inputs: [],
+        outputs: [token1],
+      });
+      mockProvider.addCallTo(pool, block, utils.FUNCTIONS_IFACE, "totalSupply", {
+        inputs: [],
+        outputs: [supply],
+      });
+
+      const [valid, , ,] = await testFetcher.getPoolData(block, pool);
```
Comment (e1Ru1o) [inline Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/pool.fetcher.spec.ts:87]: why checking only `valid` here?

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/agent.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+export const provideHandleTransaction =
+  (fetcher: PoolFetcher, poolSupplyThreshold: BigNumber, amountThresholdPercentage: BigNumber): HandleTransaction =>
+  async (txEvent: TransactionEvent): Promise<Finding[]> => {
+    const findings: Finding[] = [];
+    const block: number = txEvent.blockNumber;
+    const logs: LogDescription[] = txEvent.filterLog(utils.EVENTS_ABI);
+
+    await Promise.all(
+      logs.map(async (log) => {
+        try {
```
Comment (e1Ru1o) [inline Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/agent.ts:15]: `try/catch` not needed here

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/pool.fetcher.spec.ts (hunk 3 above)
Comment (Vxatz) [reply Apeswap-Bots/Large-LP-Deposits-Withdrawals/src/pool.fetcher.spec.ts]: Fixed, not sure what I was thinking here

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Large Apeswap LP Deposit/Withdrawal Bot
+
+## Description
+
+This bot detects LP deposits and withdrawals in Apeswap pools when the `amount0`/`amount1` and the pool's `totalSupply` are high.
+> The thresholds `AMOUNT_THRESHOLD_PERCENTAGE`, `POOL_SUPPLY_THRESHOLD` of what is considered high, can be adjusted in **src/utils.ts**.
+
+> To switch to Polygon network, set `onPolygon` to `true` in **src/utils.ts**.
```
Comment (e1Ru1o) [inline Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md:8]: Let's avoid this. Use the `initialize` function to detect the network and select the addresses accordingly.

**Review by e1Ru1o — CHANGES_REQUESTED** [review]

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md [hunk 6]
```diff
@@ -8,7 +8,7 @@ This bot detects LP deposits and withdrawals in Apeswap pools when the `amount0`
 
 ## Supported Chains
 
-- Binance Smart Chain
+- BNB Smart Chain
```
Comment (e1Ru1o) [inline Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md:11]: BNB is the token, if you want to use the acronym please use BSC.
- Change to BSC
- or keep Binance Smart Chain as before.

**Review by e1Ru1o — APPROVED** [review]
LGTM, I only have one comment respect to the README

File: Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md (hunk 6 above)
Comment (Vxatz) [reply Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md:11]: I had it the way you suggest initially, but I changed it after reading [this](https://www.binance.com/en/support/announcement/854415cf3d214371a7b60cf01ead0918), can you have a look @e1Ru1o ?

Comment (e1Ru1o) [reply Apeswap-Bots/Large-LP-Deposits-Withdrawals/README.md:11]: Interesting, keep it that way then

**Pull request merged**